src/DidCache.ts

```diff
@@ -1,16 +1,338 @@
+ *
+ * With this usage, the operation hash serves two roles (1) an identifier for an operation
+ * (2) an identifier for the DID document produced by the operation. In the code below,
+ * we always use VersionId in places where we mean (2) and an OperationHash defined below
+ * when we mean (1).
  */
-export interface DidCache {
-  /** The transaction that was COMPLETELY processed. */
-  readonly lastProcessedTransaction?: Transaction;
+export type VersionId = string;
+
+/**
+ * Alias OperationHash to string - see comment above
+ */
+export type OperationHash = string;
+
+/**
+ * Function type that updates a Did document given an operation. This would be instantiated
+ * with a function that implements json patch application. Using the interface instead of
+ * the actual function hides details of json patching from Did cache.
+ */
+export interface DidDocumentUpdater {
+  (didDoc: DidDocument, operation: WriteOperation): DidDocument;
+}
+
+/**
+ * Function type that creates a Did document given a create/recover operation.
+ */
+export interface DidDocumentCreator {
+  (createOp: WriteOperation): DidDocument;
+}
+
+/**
+ * The timestamp of an operation. We define a linear ordering of
+ * timestamps using the function lesser below.
+ */
+interface OperationTimestamp {
+  readonly transactionNumber: number;
+  readonly operationIndex: number;
+}
+
+function lesser (ts1: OperationTimestamp, ts2: OperationTimestamp): boolean {
+  return ((ts1.transactionNumber < ts2.transactionNumber) ||
+          (ts1.transactionNumber === ts2.transactionNumber) && (ts1.operationIndex < ts2.operationIndex));
+}
+
+/**
+ * Information about a write operation relevant for the DID cache, a subset of the properties exposed by
+ * WriteOperation.
+ */
+interface OperationInfo extends OperationTimestamp {
+  readonly batchFileHash: string;
+  readonly type: OperationType;
+}
+
+/**
+ * DIDCache is a singleton class whose instance holds most of the state in Sidetree node.
+ * It exposes methods to record sidetree DID state changes (create, update, delete, recover)
+ * and methods to retrieve current and historical states of a DID document.
+ *
+ * The current implementation is a main-memory implementation without any persistence. This
+ * means that when a node is powered down and restarted DID operations need to be applied
+ * from the beginning of time. This implementation will be extended in the future to support
+ * persistence.
+ */
+export class DidCache {
+  /**
+   * Map a versionId to the next versionId whenever one exists.
+   */
+  private nextVersion: Map<VersionId, VersionId> = new Map();
+
+  /**
+   * Map a operation hash to the OperationInfo which contains sufficient
+   * information to reconstruct the operation.
+   */
+  private opHashToInfo: Map<OperationHash, OperationInfo> = new Map();
+
+  /**
+   * Apply (perform) a specified DID state changing operation.
+   */
+  public apply (operation: WriteOperation): string | null {
```

Comment:
[TypeScript team guideline](https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines#null-and-undefined) suggests to use `undefined` instead of `null`, which makes sense. Apply the same guideline to all instances.

Reply:
Done.

Reply:
Fixed remaining null references in documentation and tests.